newrelic/python_agent.go

```diff
@@ -0,0 +1,67 @@
+package newrelic
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/buildpacks/libcnb"
+	"github.com/paketo-buildpacks/libpak/bard"
+	"github.com/paketo-buildpacks/libpak/sherpa"
+)
+
+type PythonAgent struct {
+	BuildpackPath   string
+	ApplicationPath string
+	Logger          bard.Logger
+}
+
+func NewPythonAgent(applicationPath string, buildpackPath string) PythonAgent {
+	return PythonAgent{
+		ApplicationPath: applicationPath,
+		BuildpackPath:   buildpackPath,
+	}
+}
+
+func (p PythonAgent) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
+	e := filepath.Walk(filepath.Join(layer.Path, "/layers/paketo-buildpacks_pip-install/packages/"), func(path string, info os.FileInfo, err error) error {
```

Comment:
This seems a tad fragile, if the pip-install buildpack changes it would impact this buildpack. 

I'm wondering if there's a way we could fetch this patch dynamically. Perhaps read `PYTHONPATH` and parse it? Or another env variable? Have you looked at an approach like that?

Reply:
PYTHONPATH does not contain information about pip-install layer
How do I get envs from other layers?

Reply:
Based on what's [here](https://github.com/paketo-buildpacks/pip#behavior) it should:

>Prepends the pip layer to the PYTHONPATH

So you should be able to read the `PYTHONPATH` env variable during build time, split on the path separator and pick out the path to the pip layer.

You don't have to do anything special, a `os.Getenv("PYTHONPATH")` should do the trick. Then `strings.Split` and pick the first item.

Reply:
Another possible way you could approach this would be to have the buildpack run python and see if it can `import newrelic`. My understanding is that the packages should be installed by the time the New Relic buildpack runs, so if you can import then you're all set. If that fails, the user hasn't installed the package.

```
> python -c 'import newrelic'  # returns 0 exit code
```

```
> python -c 'import newrelic2'  # returns 1 exit code 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'newrelic2'
```

That is probably the least fragile way to check, IMHO.

Reply:
I am probably doing something wrong, when I get PYTHONPATH at build stage or even when running application configuration stage I get the same path back 
[/layers/paketo-buildpacks_pip/pip/lib/python3.10/site-packages /layers/paketo-buildpacks_cpython/cpython]
Does not include the pip-install layer 
If I try to execute python command to import the module, errors out saying that module does not exists 
Also trying to execute command using newrelic-admin cli that gets installed by pip, errors saying that (exec: "newrelic-admin": executable file not found in $PATH)
If I change the running dir to the path for pip-install layer 
		Command: "python",
		Args:    []string{"-c", "import newrelic.agent"},
		Dir:     "/layers/paketo-buildpacks_pip-install/packages/lib/python3.10/site-packages",
It works. But this is the same, the path is hardcoded and it's what we trying to avoid.
That said from my lack of experience here, I think that file.walk on /layers/ (not has previously done in specific layer pip-install) and search for /newrelic/admin/ is best course action, this ensures that even if there is change to pip buildpack, it will not impact this 

This will work fine, until we need to test it, the root buildpack is read-only, so we can't create dummy folder in /layers/xxxx.

Let me know what you think

Reply:
I took a look at this and something is off.

When I do a build, the `pip-install` buildpack outputs:

```
  Configuring build environment
    PYTHONPATH -> "/layers/paketo-buildpacks_pip-install/packages/lib/python3.10/site-packages:$PYTHONPATH"
```

Which says it's modifying the PYTHONPATH during the build, but when I dump the env variables PYTHONPATH does not include the pip-install layer.

```
CNB_BUILDPACK_DIR=/cnb/buildpacks/paketo-buildpacks_new-relic/v5.1.0-48-g7f3ec47
SERVICE_BINDING_ROOT=/bindings
HOME=/home/cnb
LIBRARY_PATH=/layers/paketo-buildpacks_pip/pip/lib:/layers/paketo-buildpacks_cpython/cpython/lib
CNB_STACK_ID=io.buildpacks.stacks.bionic
PYTHONPATH=/layers/paketo-buildpacks_pip/pip/lib/python3.10/site-packages:/layers/paketo-buildpacks_cpython/cpython
CPATH=/layers/paketo-buildpacks_cpython/cpython/include
LD_LIBRARY_PATH=/layers/paketo-buildpacks_pip/pip/lib:/layers/paketo-buildpacks_cpython/cpython/lib
PATH=/layers/paketo-buildpacks_pip/pip/bin:/layers/paketo-buildpacks_cpython/cpython/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=f21b88f6771a
```

That is why this isn't working, but from what I can tell. It should be modifying PYTHONPATH so we should be able to do this. I need to talk with someone on the Python buildpack team and see if they have any thoughts on what's happening here. I'm not super familiar with that buildpack, so I could be missing something that's necessary here.

Reply:
OK, I think I have this working. Submitting some suggestions that should get us into a good place.

Reply:
All done, committed changes requested